src/extra/components/dld.py

```diff
@@ -0,0 +1,304 @@
+            # Missing one or more trains in the pulse information.
+            raise ValueError('missing pulse information for one or more '
+                             'trains with data')
+
+        if len(entry_counts) < min(len(data_counts), len(pulse_counts)):
+            # Missing counts in one or more trains in data, select data
+            # and pulses down.
+            from extra_data import by_id
+            train_sel = by_id[entry_counts.index.to_numpy()]
+
+            pulses = self._pulses.select_trains(train_sel)
+            kd = kd.select_trains(train_sel)
+        else:
+            pulses = self._pulses
+
+        pulse_index = pulses.build_pulse_index(pulse_dim)
+        data = kd.ndarray()
+
+        if split_axis is None:
+            # Separation disabled, insert a virtual axis in the front.
+            data = data[None, :]
+            axis = 0
+        else:
+            # Separation enabled.
+            axis = split_axis
+
+        res = []
+
+        for i in range(data.shape[axis]):
+            # Part of data to be handled in this iteration.
+            subdata = np.take(data, i, axis=axis)
+
+            num_rows = subdata.shape[1]
+            raw = subdata.ravel()
+
+            # Obtain mask which entries are actually filled.
+            if data.dtype.fields is not None:
+                field = {name for name, (dtype, offset)
+                         in data.dtype.fields.items()
+                         if np.issubdtype(dtype, np.floating)}.pop()
+
+                finite_mask = np.isfinite(raw[field])
+            elif np.issubdtype(data.dtype, np.floating):
+                finite_mask = np.isfinite(raw)
+            else:
+                raise TypeError(data.dtype)
+
+            # Obtain pulse index and repeat each row by the number of
+            # actual data in each entry, then convert to a dataframe.
+            index_df = pulse_index.repeat(
+                finite_mask.reshape(-1, num_rows).sum(axis=1)).to_frame()
+
+            # Insert additional index level enumerating actual entries.
+            if entry_level is not None:
+                index_df[entry_level] = np.flatnonzero(finite_mask) % num_rows
+
+            res.append(pd_cls(
+                raw[finite_mask], pd.MultiIndex.from_frame(index_df)))
+
+        return res if split_axis is not None else res[0]
+
+    @property
+    def detector_name(self):
+        return self._detector_name
+
+    @property
+    def control_source(self):
+        """Control source."""
+
+        if self._control_src is None:
+            raise ValueError('component is initialized with earlier data '
+                             'lacking a control source')
+
+        return self._control_src
+
+    @property
+    def instrument_source(self):
+        """Instrument source."""
+        return self._instrument_src
+
+    @property
+    def rec_params(self):
+        """Reconstruction parameters."""
+        return self.control_source.run_values()
```

Comment:
I just noticed that for REMI, you're writing the `RUN` values without the value & timestamp pairs that we normally see in RUN or CONTROL data. Whereas for the JUNGFRAU ROI reductions, we put the parameters in RUN using your `DataFile` API, which does create value & timestamp datasets.

Do we consider that a bug? If we do and we want to fix it, we might pre-emptively account for the difference here, i.e. discard the `.timestamp` keys and remove the `.value` suffix from the others.

(The mockdata framework will create timestamp & value pairs, so that's also not fully representative of the normal files)

Reply:
Yes, this is definitely a bug. The `.timestamp` part quite often eludes me, as I rarely if ever used it in the field. Thanks for spotting.

Reply:
Would you then drop `.value` or *simulate* it? I gravitate towards the former (and I wanted to add a keyword argument to `.run_values()` for a while now to facilitate that btw).

Reply:
I made up my mind, since it doesn't openly imply it comes from `RUN` and dating it is kind of pointless (you'd want the modification date of the underlying YAML file now, whatever it is later).